File: common/src/main/kotlin/net/impleri/playerskills/events/handlers/InternalEvents.kt
package net.impleri.playerskills.events.handlers

import dev.architectury.event.events.common.PlayerEvent
import dev.architectury.registry.ReloadListenerRegistry
import net.impleri.playerskills.api.BlockRestrictions
import net.impleri.playerskills.data.ItemRestrictionDataLoader
import net.impleri.playerskills.data.MobRestrictionDataLoader
import net.impleri.playerskills.data.SkillsDataLoader
import net.impleri.playerskills.events.SkillChangedEvent
import net.impleri.playerskills.server.NetHandler
import net.impleri.playerskills.skills.registry.Players
import net.minecraft.server.MinecraftServer
import net.minecraft.server.level.ServerPlayer
import net.minecraft.server.packs.PackType
import net.minecraft.server.packs.resources.ResourceManager
import net.minecraft.server.packs.resources.ResourceManagerReloadListener
import java.util.UUID

internal class InternalEvents : ResourceManagerReloadListener {
  fun registerEvents() {
    // Player Skills Events
    SkillChangedEvent.EVENT.register { onSkillChanged(it) }

    PlayerEvent.PLAYER_JOIN.register { onJoin(it) }
    PlayerEvent.PLAYER_QUIT.register { onQuit(it) }

    // Vanilla Events
    ReloadListenerRegistry.register(PackType.SERVER_DATA, this)
    ReloadListenerRegistry.register(PackType.SERVER_DATA, SkillsDataLoader())
    ReloadListenerRegistry.register(PackType.SERVER_DATA, ItemRestrictionDataLoader())
    ReloadListenerRegistry.register(PackType.SERVER_DATA, MobRestrictionDataLoader())
  }

  override fun onResourceManagerReload(resourceManager: ResourceManager) {
    val players = Players.close()
    Players.open(players)

    if (EventHandlers.server.isInitialized()) {
      EventHandlers.server.value.playerList
        .players
        .forEach { player -> NetHandler.syncPlayer(player) }
    }
  }

  fun resyncPlayer(server: Lazy<MinecraftServer>, playerId: UUID) {
    if (server.isInitialized()) {
      server.value.playerList.getPlayer(playerId)?.let { NetHandler.syncPlayer(it) }
    }
  }

  private val playerMap = HashMap<ServerPlayer, Long>()

  private fun onJoin(player: ServerPlayer) {
    playerMap[player] = BlockRestrictions.getReplacementsCountFor(player)
  }

  private fun onQuit(player: ServerPlayer) {
    playerMap.remove(player)
  }

  private fun onSkillChanged(event: SkillChangedEvent<*>) {
    NetHandler.syncPlayer(event)

    // get the ServerPlayer we should have from the initial filling of the cache
    val eventPlayerId = event.player.uuid

    playerMap.keys.firstOrNull { eventPlayerId.equals(it.uuid) }
      ?.let {
        val originalCount = playerMap[it]
        val newCount = BlockRestrictions.getReplacementsCountFor(it)

        // We're assuming that the number of replaced blocks should change if a skill change actually changes replacements
        // If we run into an issue where a skills change should trigger a refresh but the count difference doesn't change,
        // we'll have to rework this
        if (originalCount == newCount) {
          return
        }

        playerMap[it] = newCount
        it.getLevel().chunkSource.chunkMap.updatePlayerStatus(it, true)
      }
  }
}

File: common/src/main/scala/net/impleri/playerskills/events/handlers/InternalEvents.scala
package net.impleri.playerskills.events.handlers

import dev.architectury.registry.ReloadListenerRegistry
import net.impleri.playerskills.data.SkillsDataLoader
import net.impleri.playerskills.events.SkillChangedEvent
import net.impleri.playerskills.server.NetHandler
import net.minecraft.server.MinecraftServer
import net.minecraft.server.packs.PackType
import net.minecraft.server.packs.resources.ResourceManager
import net.minecraft.server.packs.resources.ResourceManagerReloadListener

import java.util.UUID
import scala.util.chaining.scalaUtilChainingOps

//trait BlockSync {
//    private var playerMap: mutable.HashMap[ServerPlayer, Long] = mutable.HashMap()
////  private def onJoin(player: ServerPlayer): Unit = playerMap.update(player, BlockRestrictions.getReplacementsCountFor(player))
////  private def onQuit(player: ServerPlayer) = playerMap.remove(player)
//
//  protected def checkForBlockUpdates(player: ServerPlayer, prev: Long): Boolean = {
//    val next = BlockRestrictions.getReplacementsCountFor(prev)
//
//    // We're assuming that the number of replaced blocks should change if a skill change actually changes replacements
//    // If we run into an issue where a skills change should trigger a refresh but the count difference doesn't change,
//    // we'll have to rework this
//    if (prev != next) {
//      playerMap.update(player, next)
//      true
//    } else false
//      }
//
//  protected def maybeUpdateBlocks(playerId: UUID): Unit = {
//    playerMap.find(_._1.getUUID == playerId)
//      .flatMap(t => if (checkForBlockUpdates(t._1, t._2)) Some(t._1) else None)
//      .foreach(p => p.getLevel.chunkSource.chunkMap.updatePlayerStatus(p, true))
//  }
//}

case class InternalEvents(onReload: ResourceManager => Unit) extends ResourceManagerReloadListener {
  private[handlers] def registerEvents(): Unit = {
    // Player Skills Events
    SkillChangedEvent.EVENT.register(onSkillChanged _)

    //    PlayerEvent.PLAYER_JOIN.register(onJoin _)
    //    PlayerEvent.PLAYER_QUIT.register(onQuit _ )

    // Vanilla Events
    ReloadListenerRegistry.register(PackType.SERVER_DATA, this)
    ReloadListenerRegistry.register(PackType.SERVER_DATA, SkillsDataLoader())
  }

  override def onResourceManagerReload(resourceManager: ResourceManager): Unit = onReload(resourceManager)

  def resyncPlayer(server: MinecraftServer, playerId: UUID): Unit = {
    server.getPlayerList.getPlayer(playerId)
      .tap(NetHandler.syncPlayer(_))
  }

  private def onSkillChanged(event: SkillChangedEvent[_]): Unit = {
    NetHandler.syncPlayer(event)
    //    maybeUpdateBlocks(event.player.getUUID)
  }
}
